_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+
+
+>This means that `last_updated` indeed does not fall behind more than the execution time of `auto_replenish` plus a constant dependent on the bucket configuration.
+
+
+Here, the “constant dependent on the bucket configuration” was `refill_time/size` , rounded down. This is indeed implies our above specified property, so when we revisited `auto_replenish` a few months later to add the following two `debug_asserts!` derived from our formal property
+
+```rs
+// time_adjustment = tokens * (refill_time / size)
+debug_assert!((now - last_update) >= time_adjustment);
+// inequality slightly rewritten to avoid division
+debug_assert!((now - last_update - time_adjustment) * size < refill_time);
+```
+
+we expected the verification to succeed. The “VERIFICATION FAILED” output Kani presented us with was unexpected to say the least.
+
+So what went wrong? In the hand-written proof, the error was assuming that `-⌊-x⌋ = ⌊x⌋` (had this step been gotten correctly, the bound would have been `refill_time/size` rounded *up*, which obviously allows for violations). To see how our code actually violates the property, we need to have a look at how the relevant part of `auto_replenish` was actually implemented:
+
+```rs
+let time_delta = self.last_update.elapsed().as_nanos() as u64;
+
+// tokens = time_delta / (refill_time / size) rewritten to not run into
+// integer division issues
+let tokens = (time_delta * self.size) / self.refill_time;
+let time_adjustment = (tokens * self.refill_time) / self.size
+
+self.last_update += Duration::from_nanos(time_adjustment);
+```
+
+The issue lies in the way we compute `time_adjustment`: Consider a bucket of size 2 with refill time 3ns and assume a time delta of 11ns. We compute `11⋅2/3 ≈ 7` tokens, and then a time adjustment of `7⋅3/2 ≈ 10ns`. However, 10ns  is only enough to replenish `10⋅2/3 ≈ 6`  tokens! The problem here is that 7 tokens do not take an integer number of nanoseconds to replenish. They take 10.5ns. However the integer division rounds this down, and thus the guest essentially gets to use those 0.5ns twice. Assuming the guest can time when it triggers down to nanosecond precision, and the rate limiter is configured such that `refill_time/size` is not an integer, the guest could theoretically cause these fractional nanosecond to accumulate to get an extra token every `10⁶ ⋅ refill_time/size ⋅ max(1, refill_time/size)` nanoseconds. For a rate limiter configured at 1GB/s, this would be an excess of 1KB/s.
```

Comment:
```suggestion
The issue lies in the way we compute `time_adjustment`: Consider a bucket of size 2 with refill time 3ns and assume a time delta of 11ns. We compute `11⋅2/3 ≈ 7` tokens, and then a time adjustment of `7⋅3/2 ≈ 10ns`. However, 10ns  is only enough to replenish `10⋅2/3 ≈ 6`  tokens! The problem here is that 7 tokens do not take an integer number of nanoseconds to replenish. They take 10.5ns. However the integer division rounds this down, and thus the guest essentially gets to use those 0.5ns twice. Assuming the guest can time when it triggers down to nanosecond precision, and the rate limiter is configured such that `refill_time/size` is not an integer, the guest could theoretically cause these fractional nanosecond to accumulate to get an extra token every `10⁶ ⋅ refill_time/size ⋅ max(1, refill_time/size)` nanoseconds. **For a rate limiter configured at 1GB/s, this would be an excess of 1KB/s.**
```